west/cmd/project.py

```diff
@@ -0,0 +1,604 @@
+    # the default path uses it).
+
+    if args.manifest:
+        return args.manifest
+
+    if not args.zephyr_base:
+        log.die('Explicit path to manifest file not given, and Zephyr '
+                'base directory not specified (via --zephyr-base or '
+                '$ZEPHYR_BASE)')
+
+    return os.path.join(args.zephyr_base, 'manifest', 'manifest.yml')
+
+
+def _fetch(project):
+    # Clone 'project' if it does not exist, and 'git fetch' it otherwise. Also
+    # update the 'manifest-rev' branch to point to the revision specified in
+    # the manifest.
+    #
+    # Returns True if the project already existed, and False if it was cloned.
+
+    if os.path.exists(project.path):
+        existed = True
+        _verify_repo(project)
+        _inf(project, '=== Fetching changes for (name-and-path) ===')
+        _git(project, 'remote update')
+    else:
+        existed = False
+        _inf(project, '=== Cloning (name-and-path) ===')
+        _git_base(project, 'clone (url)/(name) (path)')
+
+    # Update manifest-rev branch
+    _git(project,
+         'update-ref refs/heads/(manifest-rev-branch) {}'.format(
+             (project.revision if _is_sha(project.revision) else
+                 'origin/' + project.revision)))
+
+    return existed
+
+
+def _rebase(project):
+    _inf(project, '=== Rebasing (name-and-path) to (manifest-rev-branch) ===')
+    _git(project, 'rebase (manifest-rev-branch)')
+
+
+def _cloned(project):
+    # Returns True if the project's repository exists and looks like a Git
+    # repository.
+    #
+    # Prints a warning if the project's clone path exist but doesn't look like
+    # a Git repository.
+
+    if os.path.exists(project.path):
+        _verify_repo(project)
+        return True
+
+    return False
+
+
+def _verify_repo(project):
+    # Raises an error if the project's clone path is not the top-level
+    # directory of a Git repository
+
+    # --is-inside-work-tree doesn't require that the directory is the top-level
+    # directory of the Git repository. Use --show-cdup instead, which prints an
+    # empty string (i.e., just a newline, which we strip) in that case.
+    res = _git(project, 'rev-parse --show-cdup', capture_stdout=True,
+               check=False)
+
+    if res.returncode != 0 or res.stdout != "":
+        _die(project, '(name-and-path) is not the top-level directory of a '
+                      'Git repository, as reported by '
+                      "'git rev-parse --show-cdup'")
+
+
+def _has_branch(project, branch):
+    return _git(project, 'show-ref --quiet --verify refs/heads/' + branch,
+                check=False).returncode == 0
+
+
+def _is_sha(s):
+    try:
+        int(s, 16)
+    except ValueError:
+        return False
+
+    return len(s) == 40
+
+
+def _git_base(project, cmd, *, extra_args=(), capture_stdout=False, check=True):
+    # Runs a git command in the base directory. See _git_helper() for parameter
```

Comment:
Still not sure what "base directory" means here, but still I think you need to be going back up the tree until you find WEST_DIR. Am I missing something?

Reply:
Means the current directory (as opposed to one of the project directories). I was assuming the project commands would just do their thing in whatever directory you happen to be in.

Does `repo` do a search?

Reply:
I need to try this out, but wouldn't fetch need to go back to the top level directory to check paths properly?

Reply:
Was thinking plain git commands. Maybe it works differently in repo though.

Reply:
So I tried this out, and I think it's looking for ZEPHYR_BASE when it need not:

```
$ west init -w https://github.com/ulfalizer/west --wr project-rfc test
$ cd test/
$ ls
manifest  west
$ west fetch
warning: --zephyr-base missing and no ZEPHYR_BASE in the environment
fatal error: Explicit path to manifest file not given, and Zephyr base directory not specified (via --zephyr-base or $ZEPHYR_BASE)
```

If ZEPHYR_BASE continues to mean the same thing as always, that's not where the manifest is: the manifest is the 'manifest' directory relative to `.west` (which could be in $PWD or somewhere above it).

Reply:
Note that we  put the west repo in the top level directory (where `.west` is, called WEST_DIR in the bootstrap script), but that's something we could move around if we wanted since nobody's using it yet. The marker of the top level is the `.west` dir, analogous to `.git`.